lib/experimental/sync/class-gutenberg-http-signaling-server.php

```diff
@@ -0,0 +1,343 @@
+			if ( ! $message ) {
+				die( 'no message' );
+			}
+			static::handle_message_operation( $message );
+		}
+
+		static::clean_up_old_connections();
+		exit;
+	}
+
+	/**
+	 * Initializes the paths of the temporary files used.
+	 */
+	private static function initialize_paths() {
+		static::$subscriber_to_messages_path = get_temp_dir() . DIRECTORY_SEPARATOR . 'subscribers_to_messages.txt';
+		// Example: array( 2323232121 => array( 'message hello','handshake message' ) ).
+
+		static::$topics_to_subscribers_path = get_temp_dir() . DIRECTORY_SEPARATOR . 'topics_to_subscribers.txt';
+		// Example: array( 'doc1: array( 2323232121 ), 'doc2: array( 2323232123, 2323232121 ) ).
+
+		static::$subscribers_to_last_connection_path = get_temp_dir() . DIRECTORY_SEPARATOR . 'subscribers_to_last_connection.txt';
+		// Example: array( 2323232121 => 34343433323(timestamp) ).
+	}
+
+	/**
+	 * Reads the contents of $fd and returns them unserialized.
+	 *
+	 * @access private
+	 * @internal
+	 *
+	 * @param resource $fd  A file descriptor.
+	 *
+	 * @return array Unserialized contents of fd.
+	 */
+	private static function get_contents_from_file_descriptor( $fd ) {
+		$contents_raw = stream_get_contents( $fd );
+		$result       = array();
+		if ( $contents_raw ) {
+			$result = unserialize( $contents_raw );
+		}
+		return $result;
+	}
+
+
+	/**
+	 * Makes the file descriptor content of $fd equal to the serialization of content.
+	 * Overwrites what was previously in $fd.
+	 *
+	 * @access private
+	 * @internal
+	 *
+	 * @param resource $fd      A file descriptor.
+	 * @param array    $content Content to be serialized and written in a file descriptor.
+	 */
+	private static function save_contents_to_file_descriptor( $fd, $content ) {
+		rewind( $fd );
+		$data = serialize( $content );
+		fwrite( $fd, $data );
+		ftruncate( $fd, strlen( $data ) );
+	}
+
+	/**
+	 * Handles a wp_ajax signaling server request of client that wants to retrieve its messages.
+	 *
+	 * It returns the client a response following the
+	 * {@link https://developer.mozilla.org/en-US/docs/Web/API/Server-sent_events/Using_server-sent_events#event_stream_format Event stream format}.
+	 *
+	 * ```
+	 * id: <Event ID>
+	 * retry: <Reconnection time, in ms>
+	 * event: <The type of event>
+	 * data: <The message to be sent>
+	 * ```
+	 */
+	private static function handle_message_read_request() {
+		header( 'Content-Type: text/event-stream' );
+		header( 'Cache-Control: no-cache' );
+		echo 'retry: 3000' . PHP_EOL;
+		$fd = fopen( static::$subscriber_to_messages_path, 'c+' );
+		if( ! $fd ) {
+			die( 'Could not open required file.' );
+		}
+		flock( $fd, LOCK_EX );
+		$subscriber_to_messages = static::get_contents_from_file_descriptor( $fd );
+		if ( isset( $subscriber_to_messages[ static::$subscriber_id ] ) && count( $subscriber_to_messages[ static::$subscriber_id ] ) > 0 ) {
+			echo 'id: ' . time() . PHP_EOL;
+			echo 'event: message' . PHP_EOL;
```

Comment:
It sounds like the [Event stream format](https://developer.mozilla.org/en-US/docs/Web/API/Server-sent_events/Using_server-sent_events#event_stream_format) allows us to send named/unnamed events: do you have any thoughts about this?

Reply:
I thought about using an event specific to messages, and for now, all our events are messages. I guess using named events leaves the door open to add different types of message events if we need them in the future.